// @midnight-ntwrk/wallet-sdk-shielded/dist/v1/TransactionHistory.js
import { CoreWallet } from './CoreWallet.js';
export const makeDefaultTransactionHistoryCapability = () => {
    return {
        updateTxHistory: (state, newTxs) => {
            return newTxs.reduce((acc, tx) => CoreWallet.addTransaction(acc, tx), state);
        },
        transactionHistory: (state) => {
            return state.txHistoryArray;
        },
        progress: (state) => {
            return {
                appliedIndex: state.progress.appliedIndex,
                highestRelevantWalletIndex: state.progress.highestRelevantWalletIndex,
                highestIndex: state.progress.highestIndex,
                highestRelevantIndex: state.progress.highestRelevantIndex,
            };
        },
    };
};
export const makeSimulatorTransactionHistoryCapability = () => {
    return {
        updateTxHistory: (state, newTxs) => {
            return CoreWallet.updateTxHistory(state, newTxs); // @TODO fix this cast
        },
        transactionHistory: (state) => {
            return state.txHistoryArray; // @TODO fix this cast
        },
        progress: (state) => {
            return {
                appliedIndex: state.progress.appliedIndex,
                highestRelevantWalletIndex: state.progress.highestRelevantWalletIndex,
                highestIndex: state.progress.highestIndex,
                highestRelevantIndex: state.progress.highestRelevantIndex,
            };
        },
    };
};
export const makeDiscardTransactionHistoryCapability = () => {
    return {
        updateTxHistory: (state) => {
            return state;
        },
        transactionHistory: (state) => {
            return state.txHistoryArray;
        },
        progress: (state) => {
            return {
                appliedIndex: state.progress.appliedIndex,
                highestRelevantWalletIndex: state.progress.highestRelevantWalletIndex,
                highestIndex: state.progress.highestIndex,
                highestRelevantIndex: state.progress.highestRelevantIndex,
            };
        },
    };
};
